File: backend/app/services/contractor_service.py
"""
Contractor service.

Provides higher-level operations for managing contractors.
"""

from app.domain.models import Contractor
from app.repositories.contractor_repo import ContractorRepository


class ContractorService:
    """Service for retrieving and managing contractor information."""

    def __init__(self, contractor_repo: ContractorRepository):
        self.contractor_repo = contractor_repo

    def list_for_user(self, user_id: int) -> list[Contractor]:
        """
        Return all contractors belonging to a specific user.
        """
        return self.contractor_repo.get_by_user(user_id)

    def get(self, contractor_id: int) -> Contractor | None:
        """
        Retrieve a single contractor by ID.
        """
        return self.contractor_repo.get_by_id(contractor_id)

    def create(self, user_id: int, name: str) -> Contractor:
        """
        Create a new contractor for the user.
        """
        return self.contractor_repo.create(user_id, name)

File: frontend/src/services/contractorService.ts
import apiClient from "./api";
import type { Contractor, ContractorCreate } from "@/types/contractor";

export const contractorService = {
  async getByUser(userId: number): Promise<Contractor[]> {
    const response = await apiClient.get<Contractor[]>(
      `/contractors/user/${userId}`,
    );
    return response.data;
  },

  async getById(contractorId: number): Promise<Contractor> {
    const response = await apiClient.get<Contractor>(
      `/contractors/${contractorId}`,
    );
    return response.data;
  },

  async createForUser(
    userId: number,
    data: ContractorCreate,
  ): Promise<Contractor> {
    const response = await apiClient.post<Contractor>(
      `/contractors/user/${userId}`,
      data,
    );
    return response.data;
  },
};
